FLT/GL2/Subgroup.lean

```diff
@@ -0,0 +1,212 @@
+import Mathlib.RingTheory.Henselian
+import Mathlib.RingTheory.Valuation.ValuationRing
+import Mathlib.RingTheory.Valuation.ValuativeRel
+import Mathlib.Topology.Algebra.Valued.ValuativeRel
+/-!
+
+# Definitions of various compact open subgrups of M₂ or GL₂
+# in the local and global setting
+
+We define U₁(v) as a subgroup of GL₂(Fᵥ), and U₁(S) as a subgroup
+of GL₂(𝔸_F^∞). We introduce the concept
+of a rigidification `r : (D ⊗[F] 𝔸_F^∞) ≅ M₂(𝔸_F^∞)` in order
+to push U₁(S) over to a subgroup of `(D ⊗[F] 𝔸_F^∞)ˣ`.
+
+-/
+
+
+/-
+
+If I just have a ValuativeRel, do I get a ring of integers and a residue field?
+
+-/
+open scoped ValuativeRel
+
+variable (K : Type*) [Field K] [ValuativeRel K]
+
+--#check (𝒪[K])
+--#check (𝓀[K])
+
+/- GL_2(O_K) now easy: -/
+example : Subgroup (GL (Fin 2) K) := 𝒪[K].matrix.units
+
+-- do we need this?
+example (A : Type*) [Ring A] (R : Subring A) (n : Type*) [Fintype n] [DecidableEq n] :
+    Matrix n n R ≃+* (R.matrix : Subring (Matrix n n A)) := sorry
+
+example : 𝒪[K] →+* K := 𝒪[K].subtype
+example : 𝒪[K] →+* 𝓀[K] := IsLocalRing.residue ↥𝒪[K]
+
+namespace M2
+
+def UpperTriangular (k : Type*) [CommRing k] : Subring (Matrix (Fin 2) (Fin 2) k) where
+  carrier := {M | M.BlockTriangular id}
+  mul_mem' := by
+    intro a b ha hb i j hij
+    simp only [Matrix.mul_apply, Fin.sum_univ_two, Fin.isValue]
+    simp only [Matrix.BlockTriangular, id_eq, Set.mem_setOf_eq] at ha hb
+    simp only [id_eq] at hij
+    rw [ha (lt_of_le_of_lt (Fin.zero_le j) hij), hb (lt_of_lt_of_le hij
+      (StrictMono.maximal_preimage_top (fun ⦃a b⦄ a ↦ a) rfl i) : j < 1)]
+    ring
+  one_mem' := by intro i j hij; exact Matrix.one_apply_ne' (id (ne_of_lt hij))
+  add_mem' := by intro a b ha hb i j hij; simp [ha hij,hb hij]
+  zero_mem' := by intro i j hij; rfl
+  neg_mem' := by intro x hx i j hij; simp [hx hij]
+
+lemma mem_upperTriangular_iff {k : Type*} [CommRing k] (M : Matrix (Fin 2) (Fin 2) k) :
+    M ∈ UpperTriangular k ↔ M 1 0 = 0 := by
+  constructor
+  · intro hM
+    exact hM Nat.one_pos
+  · intro hM i j hij
+    rw [(Fin.lt_one_iff j).1 (lt_of_lt_of_le hij (StrictMono.maximal_preimage_top (fun ⦃a b⦄ a ↦ a)
+    rfl (id i))),Fin.eq_one_of_ne_zero i (ne_of_lt (lt_of_le_of_lt (Fin.zero_le j) hij)).symm]
+    exact hM
+
+def UpperTriangularToProd {k : Type*} [CommRing k] : UpperTriangular k →+* k × k where
+  toFun M := (M.1 0 0, M.1 1 1)
+  map_one' := rfl
+  map_mul' x y := by
+    simp only [Fin.isValue, Subring.coe_mul, Matrix.mul_apply, Fin.sum_univ_two, Prod.mk_mul_mk,
+      Prod.mk.injEq, add_eq_left, add_eq_right]
+    constructor
+    · simp [(mem_upperTriangular_iff y.1).1 y.2]
+    · simp [(mem_upperTriangular_iff x.1).1 x.2]
+  map_zero' := rfl
+  map_add' x y := rfl
+
+end M2
+
+namespace GL2
+
+def UpperTriangular (k : Type*) [CommRing k] : Subgroup (GL (Fin 2) k) :=
+  (M2.UpperTriangular k).units
+
+def UpperTriangularToProd {k : Type*} [CommRing k] : UpperTriangular k →* kˣ × kˣ where
```

Comment:
This should just be the composition of the units of `UpperTriangularToProd`, and the isomorphism (M x N)^* = M^* x N^*, you don't need to do all the proofs again.